contracts/src/rollup/ValidatorWallet.sol

```diff
@@ -6,24 +6,93 @@ pragma solidity ^0.8.0;
 
 import "../challenge/IChallengeManager.sol";
 import "../libraries/DelegateCallAware.sol";
+import "../libraries/IGasRefunder.sol";
+import {IRollupUserAbs} from "./IRollupLogic.sol";
 import "@openzeppelin/contracts/utils/Address.sol";
 import "@openzeppelin/contracts-upgradeable/access/OwnableUpgradeable.sol";
 
-contract ValidatorWallet is OwnableUpgradeable, DelegateCallAware {
+error BadArrayLength(uint256 expected, uint256 actual);
+error NotCallerOrOwner(address actual);
+error OnlyOwnerFunctionSig(address expected, address actual);
+
+contract ValidatorWallet is OwnableUpgradeable, DelegateCallAware, GasRefundEnabled {
     using Address for address;
 
-    function initialize() external initializer onlyDelegated {
+    /// @dev the executor is allowed to call non-admin functions and get refunded
+    address public executor;
+
+    /// @dev function signatures that can only be called by the owner
+    mapping(bytes4 => bool) public onlyOwnerFuncSigs;
+
+    modifier onlyExecutorOrOwner() {
+        if (executor != _msgSender() && owner() != _msgSender())
+            revert NotCallerOrOwner(_msgSender());
+        _;
+    }
+
+    event ExecutorTransferred(address indexed previousExecutor, address indexed newExecutor);
+
+    function setExecutor(address newExecutor) external onlyExecutorOrOwner {
+        address prevExecutor = executor;
+        executor = newExecutor;
+        emit ExecutorTransferred(prevExecutor, newExecutor);
+    }
+
+    function initialize(address _executor, address _owner) external initializer onlyDelegated {
         __Ownable_init();
+        executor = _executor;
+        emit ExecutorTransferred(address(0), _executor);
+        transferOwnership(_owner);
+
+        onlyOwnerFuncSigs[IRollupUserAbs.withdrawStakerFunds.selector] = true;
+        emit OnlyOwnerFuncSigUpdated(IRollupUserAbs.withdrawStakerFunds.selector, true);
+
+        onlyOwnerFuncSigs[IRollupUserAbs.createChallenge.selector] = true;
+        emit OnlyOwnerFuncSigUpdated(IRollupUserAbs.createChallenge.selector, true);
```

Comment:
I don't think we want to make this a default onlyOwnerFuncSigs. Keep in mind anyone can create a challenge - not just the validators in dispute.